src/buffer/out/sgrStack.hpp

```diff
@@ -0,0 +1,226 @@
+                // what is effectively an "empty" push (the subsequent pop will not change
+                // the current attributes).
+                for (size_t i = 0; i < cOptions; i++)
+                {
+                    validParts |= _GraphicsOptionToFlag(rgOptions[i]);
+                }
+            }
+
+            if (_numSgrPushes < _countof(_storedSgrAttributes))
+            {
+                _storedSgrAttributes[_numSgrPushes] = currentAttributes;
+                _validAttributes[_numSgrPushes] = validParts;
+            }
+
+            if (_numSgrPushes < c_MaxBalancedPushes)
+            {
+                _numSgrPushes++;
+            }
+        }
+
+        // Method Description:
+        // - Restores text attributes by removing from the top of the internal stack,
+        //   combining them with the supplied currentAttributes, if appropriate.
+        // Arguments:
+        // - currentAttributes - The current text attributes. If only a portion of
+        //   attributes were saved on the internal stack, then those attributes will be
+        //   combined with the currentAttributes passed in to form the return value.
+        // Return Value:
+        // - The TextAttribute that has been removed from the top of the stack, possibly
+        //   combined with currentAttributes.
+        const TextAttribute Pop(const TextAttribute& currentAttributes) noexcept
+        {
+            TextAttribute result;
+
+            if (_numSgrPushes > 0)
+            {
+                _numSgrPushes--;
+
+                if (_numSgrPushes < _countof(_storedSgrAttributes))
+                {
+                    uint32_t validParts = _validAttributes[_numSgrPushes];
+
+                    if (validParts == UINT32_MAX)
+                    {
+                        return _storedSgrAttributes[_numSgrPushes];
+                    }
+                    else
+                    {
+                        return _CombineWithCurrentAttributes(currentAttributes,
+                                                             _storedSgrAttributes[_numSgrPushes],
+                                                             validParts);
+                    }
+                }
+            }
+
+            return currentAttributes;
+        }
+
+        // Xterm allows the save stack to go ten deep, so we'll follow suit. Pushes after
+        // ten deep will still remain "balanced"--once you pop back down below ten, you'll
+        // restore the appropriate text attributes. However, if you get more than a
+        // hundred pushes deep, we'll stop counting. Why unbalance somebody doing so many
+        // pushes? Putting a bound on it allows us to provide "reset" functionality: at
+        // any given point, you can execute 101 pops and know that you've taken the stack
+        // (push count) to zero. (Then you reset text attributes, and your state is
+        // clean.)
```

Comment:
There are two problems to worry about:

1. Doing more pushes than are supported by the storage.
2. How to recover from unbalanced pushes ("garbage").

VTE has not yet implemented these sequences, but [comments in their stubs](https://gitlab.gnome.org/GNOME/vte/blob/master/src/vteseq.cc#L8491) indicate their thoughts about implementing them: their envisioned storage stack is a "ring" stack--when you do a push but the stack is full, you just drop the bottom entry off the stack to make room. The VTE approach provides a simple way to handle problem (2)--you can just do N pops, where N is the number of storage slots, and then you know that the stack is now empty--you're back to a "clean" state. However, the drawback is that if an otherwise well-behaved app (all pushes and pops are balanced) exceeds the storage (problem (1)), it will experience "bad behavior" when it tries to pop past the storage limit, leaving the current text attributes in a wrong state after the app is done.

The XTerm approach handles problem (1) more gracefully--if an otherwise well-behaved app (all pushes and pops are balanced) exceeds the storage limit, it will experience "bad behavior" only while it is currently "above" the storage limit. When it pops back below the limit, everything will be fine again, and when the app exits, it leaves the current text attributes in the proper state. But it leaves no solution for dealing with problem (2)--you have to just do "a lot" of pops, and hope that you did enough to clear the stack (to get back down to zero).

These two approaches can be balanced by using a hybrid strategy, which is exactly what I've implemented here (although I didn't yet know the VTE direction when I implemented it). The hybrid approach is to tolerate balanced pushes and pops above the storage limit N, but below a larger "sanity threshold" K. This provides a bounded number of pops that returns the stack to a clean state (K), while tolerating apps that exceed the storage limit in a better way (failing to restore attributes only while the number of pushes is above the storage limit; once below the limit, everything is good again).

So now that I've had time to think about it… I think we should prefer exactly what I've implemented here.

Reply:
I agree with your logic here. I'm happy with this implementation until proven otherwise.

Reply:
In further discussions with VTE people (Egmont), they convinced me that a floating ring stack really is a better way to go. I put some comments about it in the sgrStack.hpp header.